docs/chymyst07.md

```diff
@@ -0,0 +1,731 @@
+Therefore, the reaction is
+
+```scala
+val all_done = b[Unit,Unit]
+
+go { all_done(_, reply) + remaining(0) => reply() }
+
+```
+
+Since this reaction consumes `remaining`, it should be declared at the same reaction site as the `done + remaining => ...` reaction.
+
+The complete code is
+
+```scala
+val done = m[Unit]
+val remaining = m[Int]
+val all_done = b[Unit,Unit]
+
+site(
+  go { done() + remaining(k) if k>0 => remaining(k-1) }, // Adding a guard to be safe.
+  go { all_done(_, reply) + remaining(0) => reply() }
+)
+
+remaining(n) // Emit the molecule with value `n`,
+// which is the initial number of remaining `done()` molecules.
+
+```
+
+Adding a guard `if k>0` will prevent the first reaction from running once we consume the expected number of `done()` molecules.
+This allows the user to emit more `done()` molecules than expected without disrupting the logic.
+
+### Refactoring into a library
+
+Consider what is required in order to refactor this functionality into a library.
+We would like to have a library function call such as `make_all_done()` that creates the `all_done()` molecule for us.
+
+The library function will need to declare a new reaction site.
+The `done()` molecule is an input molecule at the new reaction site.
+Therefore, this molecule cannot be already defined before we use the library call.
+The call to `make_all_done()` will have to return _both_ a new `done()` molecule and a new `all_done()` molecule.
+The user of the library will then need to emit the `done()` molecule at the end of each job.
+
+Refactoring a piece of chemistry into a library call is done by declaring the same molecules and reactions as we did above,
+all within the scope of a function.
+The function should then return just the new molecule emitters that the library user will need to use, but no other molecule emitters:
+
+```scala
+def make_all_done(n: Int): (E, EE) = {
+  val done = m[Unit]
+  val remaining = m[Int]
+  val all_done = b[Unit,Unit]
+  
+  site(
+    go { done() + remaining(k) if k>0 => remaining(k-1) }, // Adding a guard to be safe.
+    go { all_done(_, reply) + remaining(0) => reply() }
+  )
+  
+  remaining(n)
+  
+  (done, all_done)
+}
+
+```
+
+Note that we declared the molecule types to be `E` and `EE`.
+These are the types created by the `m[Unit]` and `m[Unit,Unit]` macro calls.
+The class `E` is a refinement of the class `M[Unit]`, and `EE` is a refinement `M[Unit,Unit]`.
+The refinements are pure syntactic sugar: they are needed in order to permit the syntax `done()` and `all_done()` without a compiler warning about the missing `Unit` arguments.
+Without them, we would have to write `done(())` and `all_done(())`.
+
+## Waiting until `n` jobs are finished: blocking calls
+
+A variation on the same theme is to detect when `n` jobs are finished, given that each job will emit a _blocking_ molecule `done()` at the end.
+
+TODO
+
+## Control over a shared resource (mutex, multiplex)
+
+### Single access
```

Comment:
In this suggestion, I view the idea of serialization as a single queue or a single cashier to accept the requests that must line up. The single cashier controls access to the restricted resource (e.g. the database). Hence, the single reaction to do the doWork corresponds to a single cashier. My cashier is your access molecule.